apbs/input_file/calculate/finite_difference.py

```diff
@@ -0,0 +1,1899 @@
+                self._counts = None
+                raise TypeError(
+                    f"Vector {value} does not contain positive integers."
+                )
+            self._counts.append(elem)
+        _LOGGER.debug(f"Target grid nunmbers are {self._counts}.")
+        self.adjust_counts()
+        _LOGGER.debug(
+            f"Actual grid number are {self._counts} with level {self.levels}."
+        )
+
+    @property
+    def spacings(self) -> list:
+        """Spacings of the grid in a finite-difference Poisson-Boltzmann
+        calculation.
+
+        :returns: 3-vector of positive numbers indicating grid spacings in the
+          x-, y-, and z-directions in Å. The spacings may be different in each
+          direction.
+        :raises IndexError:  if the length of the value is not 3.
+        :raises TypeError:  if the spacings are not positive numbers.
+        :raises ValueError:  if not enough information is available to calculate
+            spacings
+        """
+        if self._spacings is None:
+            if (self._lengths is not None) and (self._counts is not None):
+                spacings = []
+                for i in range(3):
+                    space = self._lengths[i] / (self._counts[i] - 1)
+                    spacings.append(space)
+                return spacings
+            else:
+                raise ValueError(
+                    f"Cannot calculate spacings because either lengths "
+                    f"{self._lengths} or counts {self._counts} is None."
+                )
+        return self._spacings
+
+    @spacings.setter
+    def spacings(self, value):
+        if len(value) != 3:
+            raise IndexError(f"Spacings are {len(value)} rather than 3.")
+        self._spacings = []
+        for i, elem in enumerate(value):
+            if not check.is_positive_definite(elem):
+                raise TypeError(
+                    f"Vector {value} does not contain positive numbers"
+                )
+            self._spacings.append(elem)
+
+    @property
+    def lengths(self) -> list:
+        """Length of the grid in a finite-difference Poisson-Boltzmann
+        calculation.
+
+        :returns: 3-vector of positive numbers indicating grid lengths in the
+          x-, y-, and z-directions in Å. The lengths may be different in each
+          direction.
+        :raises IndexError:  if the length of the value is not 3.
+        :raises TypeError:  if the lengths are not positive numbers.
+        :raises ValueError:  if not enough information is available to calculate
+            lengths
+        """
+        if self._lengths is None:
+            if (self._spacings is not None) and (self._counts is not None):
+                lengths = []
+                for i in range(3):
+                    length = self._spacings[i] * (self._counts[i] - 1)
+                    lengths.append(length)
+                return lengths
```

Comment:
Should this set self.lengths before returning?

Reply:
There's a possibility of changing counts and spacings, so I thought it would be good to leave it unset.

Reply:
OK. I will ponder over what side effects that could have later.